designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,505 @@
+
+### Loss of standarized way to expose configs
+
+First, it's difficult to know how a plugin exports configurations. A configuration can be an object or an array, and it may be accessible via a `configs` key or via a module specifier. Further, typescript-eslint actually [recommends](https://typescript-eslint.io/packages/typescript-eslint/#flat-config-extends) the use of their own utility to create configs, which [confuses users](https://github.com/typescript-eslint/typescript-eslint/issues/8496). That means you can have multiple ways of importing a config into your own. Here's an example:
+
+```js
+import js from "@eslint/js";
+import tailwind from "eslint-plugin-tailwindcss";
+import reactPlugin from "eslint-plugin-react";
+import eslintPluginImportX from 'eslint-plugin-import-x'
+
+export default [
+    js.configs.recommended,
+    ...tailwind.configs["flat/recommended"],
+    ...reactPlugin.configs.flat.recommended,
+    eslintPluginImportX.flatConfigs.recommended,
+];
+```
+
+### Confusion over `...` syntax
+
+As discussed in the previous section, configs may be objects or arrays, and the latter requires the use of the spread syntax (`...`) to include its contents into the overall array. We've found that JavaScript beginners don't know or understand spread syntax, making this distinction between objects and arrays even more confusing. 
+
+### Difficult extending configs for specific file patterns
+
+Further adding to the confusion over objects and arrays is how to take an existing config and apply it to a subset of files. This explicitly requires you to know if something is an object or array because how you extend it varies depending on the type. For objects, you do this:
+
+```js
+// eslint.config.js
+import js from "@eslint/js";
+
+export default [
+    {
+        ...js.configs.recommended,
+        files: ["**/src/safe/*.js"]
+    }
+];
+```
+
+For arrays you do this:
+
+```js
+// eslint.config.js
+import exampleConfigs from "eslint-config-example";
+
+export default [
+    ...exampleConfigs.map(config => ({
+        ...config,
+        files: ["**/src/safe/*.js"]
+    })),
+
+    // your modifications
+    {
+        rules: {
+            "no-unused-vars": "warn"
+        }
+    }
+];
+```
+
+In both cases, users need to use the spread syntax, and in both cases, just glancing at the config object doesn't really tell you what exactly is happening. It's a lot of syntax for what used to be a single statement in eslintrc.
+
+## Goals
+
+There are two goals with this design:
+
+1. **Make it easier to configure ESLint.** For users, we want to reduce unnecessarily boilerplate and guesswork as much as possible.
+1. **Encourage plugins to use `configs`.** For plugin authors, we want to encourage the use of a consistent entrypoint, `configs`, where users can find predefined configurations. 
+
+## Detailed Design
+
+Design Summary:
+
+1. Allow arrays in config arrays (in addition to objects)
```

Comment:
Should we also support Promise in the configs array? Currently, ESLint support returning a single promise as `default exports`. If any of the shared config returns a promise, users need to either use TLA or wrapper everything with `Promise.all`, like:

```ts
export default Promise.all([
    js.configs.recommended,
    ...tailwind.configs["flat/recommended"],
    ...reactPlugin.configs.flat.recommended,
    somePromise,
]).flat()
```

which:

1. It's raise the bar to know how for users
2. Verbose to write
3. Need to undo the flatten on the user side again,

The promise can also return either an Object or an array, so I see the type might need to be:

```ts
Array<ConfigItem | Array<ConfigItem> | Promise<ConfigItem | Array<ConfigItem>>>
```

Reply:
That's beyond the scope of this RFC. If you'd like to suggest that, please open a separate issue as a starting point.

Reply:
I know this is a bit in the weeds/off topic, but TLA isn't quite necessary since you could use the `await` syntax with an async IIFE as well; the default export accepting a promise is fundamentally all you need:
```js
// CJS

export default (async () => defineConfig([
    js.configs.recommended,
    ...tailwind.configs["flat/recommended"],
    ...reactPlugin.configs.flat.recommended,
    await somePlugin.someAsyncConfig(),
    {
       extends: [(await import('someESM.mjs')).configs.foo],
       // some config
    }
]))();
```

This lets the user avoid having to do the flattening themself. Of course you'd also have to know to think of that, so maybe there's a documentation opportunity here 🤷‍♂️  but it seems reasonable to me that the config helper should be synchronous and require the user to pre-await any promises. Might also be a good opportunity for a special case error message in the helper if a promise is encountered where an array/object is expected.